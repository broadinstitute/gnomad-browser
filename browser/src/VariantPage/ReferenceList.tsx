import React from 'react'

import { ExternalLink, List, ListItem } from '@gnomad/ui'

type Props = {
  variant: {
    variant_id: string
    reference_genome: 'GRCh37' | 'GRCh38'
    chrom: string
    pos: number
    ref: string
    caid?: string
    rsids?: string[]
    clinvar?: {
      clinvar_variation_id: string
    }
  }
}

export const NcbiReference = (variantRsids: string[]) => {
  return (
    <>
      {variantRsids.length === 1 && (
        // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <ListItem>
          {/* @ts-expect-error TS(2769) FIXME: No overload matches this call. */}
          <ExternalLink href={`http://www.ncbi.nlm.nih.gov/snp/${variantRsids[0]}`}>
            dbSNP ({variantRsids[0]})
          </ExternalLink>
        </ListItem>
      )}
      {variantRsids.length > 1 && (
        // @ts-expect-error TS(2746) FIXME: This JSX tag's 'children' prop expects a single ch... Remove this comment to see the full error message
        <ListItem>
          dbSNP ({' '}
          {(
            variantRsids
              .map((rsid) => (
                // @ts-expect-error TS(2786) FIXME: 'ExternalLink' cannot be used as a JSX component.
                <ExternalLink key={rsid} href={`https://www.ncbi.nlm.nih.gov/snp/${rsid}`}>
                  {rsid}
                </ExternalLink>
              ))
              // @ts-expect-error TS(2769) FIXME: No overload matches this call.
              .reduce((acc, el) => [...acc, ', ', el], []) as any
          ).slice(1)}
          )
        </ListItem>
      )}
    </>
  )
}

export const ClinvarReference = (variantClinvarVariationId: string) => {
  return (
    // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
    <ListItem>
      {/* @ts-expect-error TS(2769) FIXME: No overload matches this call. */}
      <ExternalLink
        href={`https://www.ncbi.nlm.nih.gov/clinvar/variation/${variantClinvarVariationId}/`}
      >
        ClinVar ({variantClinvarVariationId})
      </ExternalLink>
    </ListItem>
  )
}

export const ReferenceList = ({ variant }: Props) => {
  const ucscReferenceGenomeId = variant.reference_genome === 'GRCh37' ? 'hg19' : 'hg38'
  const { chrom, pos, ref } = variant
  const ucscURL = `https://genome.ucsc.edu/cgi-bin/hgTracks?db=${ucscReferenceGenomeId}&highlight=${ucscReferenceGenomeId}.chr${chrom}%3A${pos}-${
    pos + (ref.length - 1)
  }&position=chr${chrom}%3A${pos - 25}-${pos + (ref.length - 1) + 25}`
  const allOfUsURL = `https://databrowser.researchallofus.org/variants/${variant.variant_id}`

  return (
    // @ts-expect-error TS(2746) FIXME: This JSX tag's 'children' prop expects a single ch... Remove this comment to see the full error message
    <List>
      {variant.rsids && NcbiReference(variant.rsids)}

      {/* @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message */}
      <ListItem>
        {/* @ts-expect-error TS(2786) FIXME: 'ExternalLink' cannot be used as a JSX component. */}
        <ExternalLink href={ucscURL}>UCSC</ExternalLink>
      </ListItem>

      {variant.clinvar && ClinvarReference(variant.clinvar.clinvar_variation_id)}

      {variant.caid && (
        // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <ListItem>
          {/* @ts-expect-error TS(2769) FIXME: No overload matches this call. */}
          <ExternalLink
            href={`https://reg.clinicalgenome.org/redmine/projects/registry/genboree_registry/by_canonicalid?canonicalid=${variant.caid}`}
          >
            ClinGen Allele Registry ({variant.caid})
          </ExternalLink>
        </ListItem>
      )}

      {variant.reference_genome === 'GRCh38' && (
        // @ts-expect-error TS(2745) FIXME: This JSX tag's 'children' prop expects type 'never... Remove this comment to see the full error message
        <ListItem>
          {/* @ts-expect-error TS(2786) FIXME: 'ExternalLink' cannot be used as a JSX component. */}
          <ExternalLink href={allOfUsURL}>All of Us</ExternalLink>
        </ListItem>
      )}
    </List>
  )
}
