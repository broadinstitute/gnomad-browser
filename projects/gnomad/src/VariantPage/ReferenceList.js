import PropTypes from 'prop-types'
import React from 'react'

import { ExternalLink, List, ListItem } from '@broad/ui'

export const ReferenceList = ({ variant }) => {
  /* eslint-disable prettier/prettier */
  const clinvarURL = variant.rsid && variant.rsid !== '.'
    ? `http://www.ncbi.nlm.nih.gov/clinvar?term=${variant.rsid}%5BVariant%20ID%5D`
    : `http://www.ncbi.nlm.nih.gov/clinvar?term=(${variant.chrom}%5BChromosome%5D)%20AND%20${variant.pos}%5BBase%20Position%20for%20Assembly%20GRCh37%5D`

  const dbsnpURL = `http://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=${variant.rsid}`

  const ucscURL = `http://genome.ucsc.edu/cgi-bin/hgTracks?db=hg19&highlight=hg19.chr${variant.chrom}%3A${variant.pos}-${variant.pos + (variant.ref.length - 1)}&position=chr${variant.chrom}%3A${variant.pos - 25}-${variant.pos + (variant.ref.length - 1) + 25}`
  /* eslint-enable prettier/prettier */

  return (
    <List>
      <ListItem>
        {variant.rsid && variant.rsid !== '.' ? (
          <ExternalLink href={dbsnpURL}>dbSNP</ExternalLink>
        ) : (
          'Not found in dbSNP'
        )}
      </ListItem>
      <ListItem>
        <ExternalLink href={ucscURL}>UCSC</ExternalLink>
      </ListItem>
      <ListItem>
        <ExternalLink href={clinvarURL}>ClinVar</ExternalLink>
      </ListItem>
    </List>
  )
}

ReferenceList.propTypes = {
  variant: PropTypes.shape({
    chrom: PropTypes.string.isRequired,
    pos: PropTypes.number.isRequired,
    ref: PropTypes.string.isRequired,
    rsid: PropTypes.string,
  }).isRequired,
}
