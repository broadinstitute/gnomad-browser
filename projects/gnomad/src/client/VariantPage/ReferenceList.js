import PropTypes from 'prop-types'
import React from 'react'

import { ExternalLink, List, ListItem } from '@broad/ui'

export const ReferenceList = ({ variant, clinvarVariant }) => {
  const ucscReferenceGenomeId = variant.reference_genome === 'GRCh37' ? 'hg19' : 'hg38'
  const { chrom, pos, ref } = variant
  /* eslint-disable-next-line prettier/prettier */
  const ucscURL = `https://genome.ucsc.edu/cgi-bin/hgTracks?db=${ucscReferenceGenomeId}&highlight=${ucscReferenceGenomeId}.chr${chrom}%3A${pos}-${pos + (ref.length - 1)}&position=chr${chrom}%3A${pos - 25}-${pos + (ref.length - 1) + 25}`

  return (
    <List>
      {variant.rsid && (
        <ListItem>
          <ExternalLink
            href={`https://www.ncbi.nlm.nih.gov/projects/SNP/snp_ref.cgi?rs=${variant.rsid}`}
          >
            dbSNP ({variant.rsid})
          </ExternalLink>
        </ListItem>
      )}
      <ListItem>
        <ExternalLink href={ucscURL}>UCSC</ExternalLink>
      </ListItem>
      {clinvarVariant && (
        <ListItem>
          <ExternalLink
            href={`https://www.ncbi.nlm.nih.gov/clinvar/?term=${clinvarVariant.allele_id}[alleleid]`}
          >
            ClinVar ({clinvarVariant.allele_id})
          </ExternalLink>
        </ListItem>
      )}
    </List>
  )
}

ReferenceList.propTypes = {
  variant: PropTypes.shape({
    reference_genome: PropTypes.oneOf(['GRCh37', 'GRCh38']).isRequired,
    chrom: PropTypes.string.isRequired,
    pos: PropTypes.number.isRequired,
    ref: PropTypes.string.isRequired,
    rsid: PropTypes.string,
  }).isRequired,
  clinvarVariant: PropTypes.shape({
    allele_id: PropTypes.number.isRequired,
  }),
}

ReferenceList.defaultProps = {
  clinvarVariant: undefined,
}
