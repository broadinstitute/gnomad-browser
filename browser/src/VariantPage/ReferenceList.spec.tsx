import { describe, expect, test } from '@jest/globals'
import 'jest-styled-components'

import { NcbiReference, ClinvarReference } from './ReferenceList'

describe('ReferenceLists NCBI external link', () => {
  test('is formatted as single hyperlink if the given array only contains 1 rsid', () =>
    expect(NcbiReference(['rs12345'])).toMatchSnapshot())

  test('is formatted as multiple comma seperated hyperlinks if the given array contains more than 1 rsid', () =>
    expect(NcbiReference(['rs12345', 'rs23456', 'rs34567'])).toMatchSnapshot())
})

describe('ReferenceLists ClinVar external link', () => {
  test('is formatted as single hyperlink with the given clinvar id', () =>
    expect(ClinvarReference('12345')).toMatchSnapshot())
})
