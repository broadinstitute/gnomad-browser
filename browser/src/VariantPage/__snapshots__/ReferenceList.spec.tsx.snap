// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceLists ClinVar external link is formatted as single hyperlink with the given clinvar id 1`] = `
<List__ListItem>
  <Link__ExternalLink
    href="https://www.ncbi.nlm.nih.gov/clinvar/variation/12345/"
  >
    ClinVar (
    12345
    )
  </Link__ExternalLink>
</List__ListItem>
`;

exports[`ReferenceLists NCBI external link is formatted as multiple comma seperated hyperlinks if the given array contains more than 1 rsid 1`] = `
<React.Fragment>
  <List__ListItem>
    dbSNP (
     
    <Link__ExternalLink
      href="https://www.ncbi.nlm.nih.gov/snp/rs12345"
    >
      rs12345
    </Link__ExternalLink>
    , 
    <Link__ExternalLink
      href="https://www.ncbi.nlm.nih.gov/snp/rs23456"
    >
      rs23456
    </Link__ExternalLink>
    , 
    <Link__ExternalLink
      href="https://www.ncbi.nlm.nih.gov/snp/rs34567"
    >
      rs34567
    </Link__ExternalLink>
    )
  </List__ListItem>
</React.Fragment>
`;

exports[`ReferenceLists NCBI external link is formatted as single hyperlink if the given array only contains 1 rsid 1`] = `
<React.Fragment>
  <List__ListItem>
    <Link__ExternalLink
      href="http://www.ncbi.nlm.nih.gov/snp/rs12345"
    >
      dbSNP (
      rs12345
      )
    </Link__ExternalLink>
  </List__ListItem>
</React.Fragment>
`;
